core/node/via_zk_verifier/src/lib.rs

```diff
@@ -0,0 +1,266 @@
+            da_client: client,
+            indexer,
+            config,
+        })
+    }
+
+    pub async fn run(mut self, mut stop_receiver: watch::Receiver<bool>) -> anyhow::Result<()> {
+        let mut timer = tokio::time::interval(self.config.poll_interval());
+        let pool = self.pool.clone();
+
+        while !*stop_receiver.borrow_and_update() {
+            tokio::select! {
+                _ = timer.tick() => { /* continue iterations */ }
+                _ = stop_receiver.changed() => break,
+            }
+
+            let mut storage = pool.connection_tagged("via_zk_verifier").await?;
+            match self.loop_iteration(&mut storage).await {
+                Ok(()) => {}
+                Err(err) => tracing::error!("Failed to process via_zk_verifier: {err}"),
+            }
+        }
+
+        tracing::info!("Stop signal received, via_zk_verifier is shutting down");
+        Ok(())
+    }
+
+    pub async fn loop_iteration(
+        &mut self,
+        storage: &mut Connection<'_, Core>,
+    ) -> anyhow::Result<()> {
+        if let Some((l1_batch_number, mut raw_tx_id)) = storage
+            .via_votes_dal()
+            .get_first_not_executed_block()
+            .await?
+        {
+            let db_raw_tx_id = H256::from_slice(&raw_tx_id);
+            tracing::info!("New non executed block ready to be processed");
+
+            raw_tx_id.reverse();
+            let proof_txid = bytes_to_txid(&raw_tx_id).context("Failed to parse tx_id")?;
+            tracing::info!("trying to get proof_txid: {}", proof_txid);
+            let proof_msgs = self.indexer.parse_transaction(&proof_txid).await?;
+            let proof_msg = self.expect_single_msg(&proof_msgs, "ProofDAReference")?;
+
+            let proof_da = match proof_msg {
+                FullInscriptionMessage::ProofDAReference(ref a) => a,
+                _ => {
+                    tracing::error!("Expected ProofDAReference, got something else");
+                    return Ok(());
+                }
+            };
+
+            let (proof_blob, batch_tx_id) = self.process_proof_da_reference(proof_da).await?;
+
+            let batch_msgs = self.indexer.parse_transaction(&batch_tx_id).await?;
+            let batch_msg = self.expect_single_msg(&batch_msgs, "L1BatchDAReference")?;
+
+            let batch_da = match batch_msg {
+                FullInscriptionMessage::L1BatchDAReference(ref a) => a,
+                _ => {
+                    tracing::error!("Expected L1BatchDAReference, got something else");
+                    return Ok(());
+                }
+            };
+
+            let (batch_blob, batch_hash) = self.process_batch_da_reference(batch_da).await?;
+
+            let is_verified = self
+                .verify_proof(batch_hash, &proof_blob.data, &batch_blob.data)
+                .await?;
+
+            storage
+                .via_votes_dal()
+                .mark_transaction_executed(l1_batch_number, db_raw_tx_id)
```

Comment:
Not sure that we need `is_executed`, please check this discussion https://github.com/vianetwork/via-core/pull/77#discussion_r1897396847.